### PR TITLE
test(KVStore): Support Proxy's apply pattern on normal write and compact log

### PR DESCRIPTION
**Pull request by CalvinNeo**
Signed-off-by: CalvinNeo <calvinneo1995@gmail.com>

### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/5170

Problem Summary:

`MockRaftStoreProxy` will act like TiFlash Proxy:
1. Functions like `normalWrite`/`compactLog` to generate and "commit" raft entries.
2. Functions like `doApply` to apply raft entries.
3. Functions like `replay` to replay raft entries from last applied index.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Debug/MockRaftStoreProxy.cpp [hunk 1]
```diff
@@ -295,6 +326,70 @@ void MockRaftStoreProxy::unsafeInvokeForTest(std::function<void(MockRaftStorePro
...
+        }
+        else
+        {
+            auto cf_name = CFToName(cmd_cf[i]);
+            auto key = RecordKVFormat::genKey(1, keys[i], 1);
+            RegionBench::setupDelRequest(request.add_requests(), cf_name, key);
+        }
+    }
+
+    if (cond.fail_before_kvstore) return;
```
Comment (JaySon-Huang) [inline dbms/src/Debug/MockRaftStoreProxy.cpp:381]: maybe use FailPointHelper and fiu_do_on instead of passing a `FailCond` to mock the failure?

Comment (CalvinNeo) [reply dbms/src/Debug/MockRaftStoreProxy.cpp]: I am afraid that failpoints take its effects on the whole program, by giving a explicit fail condition for every write can provide more flexibility in such tests.

**Review by breezewish — APPROVED** [review]
Mostly looks good

File: dbms/src/Debug/MockRaftStoreProxy.cpp [hunk 2]
```diff
@@ -295,6 +326,185 @@ void MockRaftStoreProxy::unsafeInvokeForTest(std::function<void(MockRaftStorePro
     cb(*this);
 }
 
+void MockRaftStoreProxy::bootstrap(
+    KVStore & kvs,
+    TMTContext & tmt,
+    UInt64 region_id)
+{
+    (void)tmt;
```
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/Debug/MockRaftStoreProxy.cpp]: fixed

File: dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp [hunk 3]
```diff
@@ -65,11 +64,23 @@ class RegionKVStoreTest : public ::testing::Test
         proxy_instance = std::make_unique<MockRaftStoreProxy>();
         proxy_helper = std::make_unique<TiFlashRaftProxyHelper>(MockRaftStoreProxy::SetRaftStoreProxyFFIHelper(
             RaftStoreProxyPtr{proxy_instance.get()}));
-        proxy_instance->init(100);
-
         kvstore->restore(*path_pool, proxy_helper.get());
     }
 
+    void init(bool create_regions = true)
```
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/Storages/Transaction/tests/gtest_kvstore.cpp]: fixed

**Review by solotzg — APPROVED** [review]

**Pull request merged**